src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_ioctl.c

```diff
@@ -546,11 +546,13 @@ xocl_read_axlf_helper(struct xocl_dev *xdev, struct drm_xocl_axlf *axlf_ptr)
 		goto done;
 	}
 
-	/* Compare MEM_TOPOLOGY previous vs new. */
-	if (xdev->topology != NULL) {
-		if (!memcmp(new_topology, xdev->topology, size)) {
+	/* Compare MEM_TOPOLOGY previous vs new. Ignore this and keep disable preserve_mem if not for aws.*/
+	if (xocl_is_aws(xdev) && (xdev->topology != NULL)) {
+		// m_mem_data can be of different length but we would not compare them if topology match fails
+		if (!memcmp(new_topology, xdev->topology, size) &&
```

Comment:
+ You want to compare the size first. So that you don't accidentally read data outside of a buffer from old or new topology.